sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -381,8 +393,12 @@ object DateTimeUtils {
       }
       j += 1
     }
+    if (i == 0 && j != 4) {
+      // year should have exact four digits
+      return None
```

Comment:
if `s` is not empty, we will never get into this branch right?

To handle empty `s`, I think a better idea is changing the default value of `segments` to `[-1, -1, -1]`